File: lumen-dataset/src/dataloader.rs
use rand::seq::SliceRandom;
use super::Dataset;
use super::Batcher;

pub struct DataLoader<D, B> 
where 
    D: Dataset,
    B: Batcher<Item = D::Item>
{
    dataset: D,
    batcher: B,
    batch_size: usize,
    shuffle: bool,
} 

pub struct DataLoaderIter<'a, D, B> 
where 
    D: Dataset,
    B: Batcher<Item = D::Item>
{
    loader: &'a DataLoader<D, B>,
    cursor: usize,
    indices: Vec<usize>,
}

pub struct DataLoaderIntoIter<D, B> 
where 
    D: Dataset,
    B: Batcher<Item = D::Item>
{
    loader: DataLoader<D, B>,
    cursor: usize,
    indices: Vec<usize>,
}

impl<D, B> DataLoader<D, B> 
where 
    D: Dataset,
    B: Batcher<Item = D::Item> 
{
    pub fn new(dataset: D, batcher: B, batch_size: usize, shuffle: bool) -> Self {
        Self {
            dataset,
            batcher,
            batch_size,
            shuffle,
        }
    }

    pub fn batch_count(&self) -> usize {
        self.dataset.len() / self.batch_size
    }

    pub fn iter<'a>(&'a self) -> DataLoaderIter<'a, D, B> {
        let indices = self.get_iter_indices();
        DataLoaderIter { loader: self, cursor: 0, indices }
    }

    fn get_iter_indices(&self) -> Vec<usize> {
        let len = self.dataset.len();
        let mut indices: Vec<usize> = (0..len).collect();
        
        if self.shuffle {
            let mut rng = rand::rng();
            indices.shuffle(&mut rng);
        }

        indices
    }
}

impl<D, B> DataLoader<D, B> 
where 
    D: Dataset,
    B: Batcher<Item = D::Item> + Default
{
    pub fn from_dataset(dataset: D, batch_size: usize, shuffle: bool) -> Self {
        Self::new(dataset, B::default(), batch_size, shuffle)
    }
}

impl<D, B> Iterator for DataLoaderIntoIter<D, B> 
where 
    D: Dataset,
    B: Batcher<Item = D::Item> 
{
    type Item = B::Output;
    fn next(&mut self) -> Option<Self::Item> {
        iter_next(&self.loader, &mut self.cursor, &self.indices)
    }
}

impl<'a, D, B> Iterator for DataLoaderIter<'a, D, B> 
where 
    D: Dataset,
    B: Batcher<Item = D::Item> 
{
    type Item = B::Output;
    fn next(&mut self) -> Option<Self::Item> {
        iter_next(&self.loader, &mut self.cursor, &self.indices)
    }
}

impl<D, B> IntoIterator for DataLoader<D, B> 
where 
    D: Dataset,
    B: Batcher<Item = D::Item> 
{
    type IntoIter = DataLoaderIntoIter<D, B>;
    type Item = B::Output;
    fn into_iter(self) -> Self::IntoIter {
        let indices = self.get_iter_indices();
        DataLoaderIntoIter {
            loader: self, 
            cursor: 0,
            indices,
        } 
    }
}

fn iter_next<D, B>(loader: &DataLoader<D, B>, cursor: &mut usize, indices: &[usize]) -> Option<B::Output> 
where 
    D: Dataset,
    B: Batcher<Item = D::Item> 
{
    let begin = *cursor;
    if begin >= loader.dataset.len() {
        None
    } else {
        let end = *cursor + loader.batch_size;
        let end = end.min(loader.dataset.len());
        let mut items = vec![];
        for index in begin..end {
            let index = indices[index];
            items.push(loader.dataset.get(index).unwrap());
        }        
        *cursor += items.len();
        let batch = loader.batcher.batch(items);
        Some(batch)
    }
}

File: lumen-python/lumen/dataset/dataloader.py
from .dataset import Dataset
from .. import Tensor
from typing import List, TypeVar, Callable, Any, Generic, Tuple
import math
import random


Item = TypeVar('Item', covariant=True)
Batch = TypeVar('Batch', covariant=True)


def default_batch(items: List[Item]) -> Batch:
    assert len(items) != 0, "null batch"
    
    elem = items[0]

    if isinstance(elem, Tensor):
        return Tensor.stack(items, 0)
    
    elif isinstance(elem, (tuple, list)):
        # [(img1, lab1), (img2, lab2), (img3, lab3)] => ((img1, img2, img3), (lab1, lab2, lab3))
        transposed = zip(*items) 
        return tuple(default_batch(list(samples)) for samples in transposed)

    else:
        raise RuntimeError(f'default_batch not support {type(elem)}, you need define a batch_fn')


class DataLoader(Generic[Item, Batch]):
    def __init__(
            self, 
            dataset: Dataset[Item], 
            batch_size: int = 1,
            shuffle: bool = False,
            batch_fn: Callable[[List[Item]], Batch] = default_batch,
            drop_last: bool = False):
        self.dataset = dataset
        self.batch_size = batch_size
        self.shuffle = shuffle
        self.batch_fn = batch_fn
        self.drop_last = drop_last

    def __len__(self) -> int:
        l = len(self.dataset)
        if self.drop_last:
            return l // self.batch_size
        else:
            return math.ceil(l / self.batch_size)

    def dataset_len(self) -> int:
        return len(self.dataset)

    def __iter__(self) -> 'DataLoaderIter[Item, Batch]':
        return DataLoaderIter(self, self._get_iter_indices())

    def _get_iter_indices(self) -> List[int]:
        indices = list(range(self.dataset_len()))
        
        if self.shuffle:
            random.shuffle(indices)
            
        return indices


class DataLoaderIter(Generic[Item, Batch]):
    def __init__(self, loader: DataLoader[Item, Batch], indices: List[int]):
        self.loader = loader
        self.indices = indices
        self.cursor = 0

    def __iter__(self) -> 'DataLoaderIter[Item, Batch]':
        return self

    def __next__(self) -> Batch:
        begin = self.cursor
        total_len = self.loader.dataset_len()
        
        if begin >= total_len:
            raise StopIteration
        
        end = begin + self.loader.batch_size
        if self.loader.drop_last and end > total_len:
            raise StopIteration

        real_end = min(end, total_len)
        batch_indices = self.indices[begin:real_end]
        items = [self.loader.dataset[i] for i in batch_indices]
        
        self.cursor = end
        
        batch = self.loader.batch_fn(items)
        return batch
    